azure-iot-provisioning-devicessdk/azure/iot/provisioning/devicessdk/transport/symmetric_key_transport.py

```diff
@@ -0,0 +1,24 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import abc
+import six
+
+
+@six.add_metaclass(abc.ABCMeta)
+class SymmetricKeyTransport(object):
```

Comment:
Unclear as to the point of this abstract class.

Reply:
IIRC not all protocols support all types of authentication. MQTT does not support TPM I think for example. it'd be nice to have a way to express this in the class definition itself, i'm guessing that's what Oliva is going for.

---
In reply to: [265319178](https://github.com/Azure/azure-iot-sdk-python-preview/pull/59#discussion_r265319178) [](ancestors = 265319178)

Reply:
Does this then mean that we end up with multiple transports per protocol, inheriting from a different authentication base?

i.e. AMQP that inherits from SymmetricKeyTransport, and another AMQP that inherits from a different type of authentication?

---
In reply to: [265325807](https://github.com/Azure/azure-iot-sdk-python-preview/pull/59#discussion_r265325807) [](ancestors = 265325807,265319178)

Reply:
ideally in OO programming you'd use interface and implement multiple interfaces. not sure how you'd translate this in Python. if a class can inherit from only one ABC, then maybe the ABC is unnecessary and we just rely on duck-typing?

---
In reply to: [265328420](https://github.com/Azure/azure-iot-sdk-python-preview/pull/59#discussion_r265328420) [](ancestors = 265328420,265325807,265319178)

Reply:
We could do some kind of multiple inheritance perhaps... but that can get a bit dicey in Python, or so I've heard (in practice I've never had an issue).

My recommendation is we don't tie ourselves into a particular hierarchy of abstraction before it's necessary.

---
In reply to: [265330148](https://github.com/Azure/azure-iot-sdk-python-preview/pull/59#discussion_r265330148) [](ancestors = 265330148,265328420,265325807,265319178)

Reply:
this mqtt from provisioning should be capable of handling symmetric auth and x509 auth. Some other transports also would have sas token functionality , hence I tried to go for multiple inheritance as many transports would take on this functionality. As far as i know it is possible in python.

Reply:
As per our discussion I'd like to see this class removed until the next PR.

---
In reply to: [267658451](https://github.com/Azure/azure-iot-sdk-python-preview/pull/59#discussion_r267658451) [](ancestors = 267658451)